### PR TITLE
Dpoole/cgroup fixes

### PR DESCRIPTION
**Pull request by dpoole73**
Changes to support cgroup v2 and changes in behavior when cgroup assignment fails.

Initial testing of linux was fine but when we tried testing on ubuntu 22.04 we found that cgroup v2 was not enabled there so the cgroup assignment failed.  In the existing code we just logged this and started the process anyway.  After discussions we decided that this is wrong and we need to only run it if we are able to assign to a cgroup.

Changes are:

1. detect cgroups v1 vs v2
2. use appropriate version of the api depdning on which version we find.
3. our initial design does not work for v2 because you can't put a cgroup inside an existing one if you already belong to that so we put it at the root instead
4. change the error handling so that we kill vmwatch if cgroup assignment fails (it shows up like a normal failure, with same restart semantics)
5. add variables to allow testing in environments without cgroups being operational (WSL, docker etc)
6. update integration tests accordingly (including moving to tcp ping and `nc` for simplicity

### CODE REVIEW

File: main/vmWatch.go [hunk 1]
```diff
@@ -11,15 +12,26 @@ import (
...
+	"github.com/containerd/cgroups/v3"
+	"github.com/containerd/cgroups/v3/cgroup1"
+	"github.com/containerd/cgroups/v3/cgroup2"
 	"github.com/go-kit/kit/log"
 	"github.com/opencontainers/runtime-spec/specs-go"
 )
 
 type VMWatchStatus string
 
+// Custom errors
```
Comment (frank-pang-msft) [inline main/vmWatch.go:24]: unused?

File: main/vmWatch.go [hunk 2]
```diff
@@ -248,40 +272,70 @@ func setupVMWatchCommand(s *vmWatchSettings, hEnv HandlerEnvironment) (*exec.Cmd
...
+	// get our process and use this to determine the appropriate mount points for the cgroups
 	myPid := os.Getpid()
+	memoryLimitInBytes := int64(MaxMemoryInBytes)
 
-	p := cgroups.PidPath(myPid)
+	// check cgroups mode
+	if cgroups.Mode() == cgroups.Unified {
+		ctx.Log("event", "cgroups v2 detected")
+		// in cgroup v2, we need to set the period and quota
+		// relative to one another
```
Comment (frank-pang-msft) [inline main/vmWatch.go:284]: Is there a formula or ratio for these? Might be good to include.

File: go.mod [hunk 3]
```diff
@@ -4,7 +4,7 @@ go 1.18
 
 require (
 	github.com/Azure/azure-docker-extension v0.0.0-20160802215703-0dd2f199467d
-	github.com/containerd/cgroups v1.1.0
+	github.com/containerd/cgroups/v3 v3.0.2
```
Comment (frank-pang-msft) [inline go.mod:7]: Does this cgroups/v3 package give you v1 & v2?

Comment (frank-pang-msft) [reply go.mod:7]: nvm looks like it does

File: integration-test/test/parallel/7_vmwatch.bats [hunk 4]
```diff
@@ -314,17 +314,50 @@ teardown(){
...
+    hanlder_log="$(container_read_handler_log)"
+    echo "$handler_log"
+    echo "$output"
+    echo "$status_file"
+    [[ "$output" == *'Setup VMWatch command: /var/lib/waagent/Extension/bin/VMWatch/vmwatch_linux_amd64'* ]]
+    [[ "$output" == *'Killing VMWatch process as cgroup assigment failed'* ]]
+    [[ "$output" == *'VMWatch reached max 3 retries, sleeping for 3 hours before trying again'* ]]
+
+    verify_substatus_item "$status_file" AppHealthStatus success "Application found to be healthy"
+    verify_substatus_item "$status_file" ApplicationHealthState success Healthy
```
Comment (frank-pang-msft) [inline integration-test/test/parallel/7_vmwatch.bats:351]: Can you also add check for VMWatch substatus as errored?

**Review by frank-pang-msft — COMMENTED** [review]
Minor comments

**Review by zmyzheng — APPROVED** [review]

File: main/vmWatch.go [hunk 5]
```diff
@@ -2,6 +2,7 @@ package main
 
 import (
 	"bytes"
+	//"errors"
```
Comment (klugorosado) [inline main/vmWatch.go:5]: remove comment or import

File: main/vmWatch.go [hunk 6]
```diff
@@ -248,40 +268,71 @@ func setupVMWatchCommand(s *vmWatchSettings, hEnv HandlerEnvironment) (*exec.Cmd
...
 
-	p := cgroups.PidPath(myPid)
+	// check cgroups mode
+	if cgroups.Mode() == cgroups.Unified {
+		ctx.Log("event", "cgroups v2 detected")
+		// in cgroup v2, we need to set the period and quota relative to one another.
+		// Quota is the number of microseconds in the period that process can run
+		// Period is the length of the period in microseconds
+		period := uint64(1000000)
+		cpuQuota := int64(MaxCpuQuota * 10000)
```
Comment (klugorosado) [inline main/vmWatch.go:283]: I like what you did with cpuQuota, perhaps period can have something similar?

File: main/vmWatch.go [hunk 7]
```diff
@@ -248,40 +268,71 @@ func setupVMWatchCommand(s *vmWatchSettings, hEnv HandlerEnvironment) (*exec.Cmd
...
 
-	s := specs.LinuxResources{
-		CPU: &specs.LinuxCPU{
-			Quota: &cpuQuota,
-		},
-		Memory: &specs.LinuxMemory{
-			Limit: &memoryLimitInBytes,
-		},
-	}
+		// in cgroup v1, the interval is implied, 1000 == 1 %
```
Comment (klugorosado) [inline main/vmWatch.go:312]: 1% is represented different in v1 vs v2? 
for v1, 1000 == 1%, and for v2 10000 == 1% ?

**Review by klugorosado — APPROVED** [review]
left some comments, overall looks good.

File: main/vmWatch.go (hunk 7 above)
Comment (dpoole73) [reply main/vmWatch.go:312]: yes it is very strange but experiments show that this is how it works

File: main/vmWatch.go (hunk 6 above)
Comment (dpoole73) [reply main/vmWatch.go:283]: done

**Review by frank-pang-msft — APPROVED** [review]

**Pull request merged**